### ISSUE TITLE
support pre-training or fine-tuning schemes

### ISSUE THREAD

**Issue by bertsky** [Layout-Parser/layout-model-training#1]
Thanks for sharing your work, it's awesome!

I am eager to train this on my own materials, but they are comparatively scarce and I don't have the computational capacities to train on the whole PubLayNet from scratch myself. 

So I was wondering: What changes would be needed to continue training from your pre-trained models? Or even more elaborately, do you think it would be worthwhile trying to load an existing model, fix most of the weights, and add some additional layers to the FPN for fine-tuning?

**Comment by lolipopshock** [comment]
Sorry just saw this. Sure, let me write a tutorial or build some exemplar code for the pre-training. Basically you just need to load any pre-trained weights we've provided you, and repeat the training process.

**Comment by bertsky** [comment]
Oh, that would be great – thanks in advance!

I expect just initializing with your pre-trained models and training on new data would quickly make the model forget your large and broad initial dataset, because initial gradients will be large. Apart from layer fixation I also thought about reducing learning rate or imposing restrictive gradient clipping. But I guess I'll have to go through these experiments anyway...

**Comment by Crnagora28** [comment]
> Thanks for sharing your work, it's awesome!
> 
> I am eager to train this on my own materials, but they are comparatively scarce and I don't have the computational capacities to train on the whole PubLayNet from scratch myself.
> 
> So I was wondering: What changes would be needed to continue training from your pre-trained models? Or even more elaborately, do you think it would be worthwhile trying to load an existing model, fix most of the weights, and add some additional layers to the FPN for fine-tuning?

Same question here! And congratulations on this amazing tool.

**Comment by joshcx** [comment]
Would be great to have such a script/documentation! Thanks a lot!

**Comment by nasheedyasin** [comment]
> Oh, that would be great – thanks in advance!
> 
> I expect just initializing with your pre-trained models and training on new data would quickly make the model forget your large and broad initial dataset, because initial gradients will be large. Apart from layer fixation I also thought about reducing learning rate or imposing restrictive gradient clipping. But I guess I'll have to go through these experiments anyway...

Any updates on your experiment @bertsky? I intend to something similar and would like to learn from your experience.

**Comment by nasheedyasin** [comment]
@lolipopshock I am trying to fine tune this model on my own custom data, having different classes that what the model was trained on.

**Comment by nasheedyasin** [comment]
> It would be really great to have a short tutorial on fine tuning on a custom data-set with custom labels starting from a pretrained model. 

If I'm successful in fine tuning on a custom dataset, will definitely work towards making a tutorial of the same.

**Comment by lolipopshock** [comment]
The plan for updating the repo and creating a dedicated fine-tuning tutorial has been unintentionally delayed - will get back to this project very soon in one or two weeks and released the updates. Please stay tuned :)

**Comment by VladyslavHerasymiuk** [comment]
> The plan for updating the repo and creating a dedicated fine-tuning tutorial has been unintentionally delayed - will get back to this project very soon in one or two weeks and release the updates. Please stay tuned :)

Hi! Any updates about a fine-tuning tutorial? I'm looking forward to it!

**Comment by nasheedyasin** [comment]
> > The plan for updating the repo and creating a dedicated fine-tuning tutorial has been unintentionally delayed - will get back to this project very soon in one or two weeks and release the updates. Please stay tuned :)
> 
> 
> 
> Hi! Any updates about a fine-tuning tutorial? I'm looking forward to it!

We recently updated a bunch of stuff to make the repo more flexible, I'll work on creating tutorial as and when I'm free, usually over the weekend.

**Comment by nasheedyasin** [comment]
Hi all,
I have written a draft of the tutorial to fine tune models using this repo. Please refer this [medium draft post](https://medium.com/@nasheed.ny/auto-parse-and-understand-any-document-5d72e81b0be9).

I will close this issue when it is published and update a link to the published version of the tutorial.

**Comment by lolipopshock** [comment]
> Hi all,
> Here's a draft of the [tutorial](https://medium.com/@nasheed.ny/auto-parse-and-understand-any-document-5d72e81b0be9) to fine tune models using this repo.
> 
> I will close this issue when it is published and update a link to the published version of the tutorial.

It seems the post is still not publicly available?

**Comment by lolipopshock** [comment]
> > > Hi all,
> > 
> > 
> > > Here's a draft of the [tutorial](https://medium.com/@nasheed.ny/auto-parse-and-understand-any-document-5d72e81b0be9) to fine tune models using this repo.
> > 
> > 
> > > 
> > 
> > 
> > > I will close this issue when it is published and update a link to the published version of the tutorial.
> > 
> > 
> > It seems the post is still not publicly available?
> 
> For now, to access the draft you'll need to be logged in to your medium account, once it's published (hopefully in a maximum of 3 days), you'll be able to publicly access it without logging in.

Thanks - just took a quick look and it looks nice! Would you mind if I also include it in the layout-parser's website as a tutorial for model training in the future? We can talk about the details if you join the slack channel - thanks!

**Comment by nasheedyasin** [comment]
Would love that, I'll join the channel right away xD

**Comment by nasheedyasin** [comment]
The [tutorial](https://link.medium.com/mJdjBsNrxhb) is now live on Towards Data Science.

**Comment by lolipopshock** [comment]
See #10
